File: simulator/spells/bless.py
from ..actions.action_types import BonusAction
from ..battle_map import Map
from ..effects.combatant_effect import CombatantEffect
from ..effects.limited_duration_effect import LimitedDurationEffect
from ..misc import get_attack_factories, ROUND_HORIZON
from ..conditions import Conditions, is_affected_by_any, get_swallower
from ..spells.spell import SpellStats
from ..effects.effect import EffectType
from ..actions.actoid import ActoidFlags
from ..threat_interfaces import AttackThreatModifier
from ..factory_interfaces import ThreatModifierFactory
from itertools import combinations
import numba_functions as nf
from ..utils.roll_types import ThreatModifierType

SAVING_THROW_BONUS_MULTIPLIER = 1.25


class BlessFactory(ThreatModifierFactory):
    level = 1
    range = SpellStats.Range.FEET_30.value
    target = SpellStats.Target.THREE_CREATURES
    duration = SpellStats.Duration.MINUTE
    concentration = True
    type = SpellStats.Type.BUFF
    dc = None
    dmg_type = None

    def __init__(self, action_type, caster, resource):
        super().__init__()
        self.action_type = action_type  # QUICKENED_BLESS, BLESS
        self.combatant = caster
        self.resource = resource

    def __str__(self):
        """
        Important for FSM building
        """
        return "BlessFactory"

    def get_ability_name(self):
        return "Bless"

    def get_quickened_kwargs(self):
        return {'caster': self.combatant, 'resource': self.resource}

    def get_eligible_targets(self):
        swallower = get_swallower(self.combatant)
        if swallower:
            return [self.combatant]
        allies = Map.get().get_non_swallowed_allies_within_radius(self.combatant, BlessFactory.range)
        allies.append(self.combatant)
        return combinations(allies, 3)

    def create_all(self, previous_action_in_dag=None):
        targets = self.get_eligible_targets()
        return [Bless(t, self) for t in targets]

    def calculate_threat_to_target(self, target, **kwargs):
        """
        Calculates the threat the factory is capable of dealing to a specific target.
        This is useful for calculating threat_in from the abilities of enemies
        """
        max_threat_increase = 0
        afs = get_attack_factories(target)
        for af in afs:
            eligible_targets = af.get_eligible_targets()
            for et in eligible_targets:
                threat_inc = af.calculate_threat_to_target_delta(et, {ThreatModifierType.TO_HIT_DIE: (1, 4)})
                max_threat_increase = max(threat_inc, max_threat_increase)
        return max_threat_increase * SAVING_THROW_BONUS_MULTIPLIER * ROUND_HORIZON

    def calculate_max_threat(self):
        targets = self.get_eligible_targets()
        if not targets:
            return 0
        return max([self.calculate_threat_to_target(t) for t in targets])


class Bless(AttackThreatModifier, CombatantEffect, LimitedDurationEffect):
    def __init__(self, targets, factory):
        AttackThreatModifier.__init__(self, ActoidFlags.IS_SPELL)
        CombatantEffect.__init__(self, factory.combatant, targets)
        LimitedDurationEffect.__init__(self, factory.combatant, turns=10)
        self.factory = factory

    def __str__(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_BLESS else "") + f"Bless on {self.combatants[0]}, {self.combatants[1]} and {self.combatants[2]}"

    def shorthand_str(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_BLESS else "") + f"Bless"

    def get_effect_type(self):
        return EffectType.BLESS

    def activate(self, **kwargs):
        Map.get().effect_tracker.add(self)
        # todo should check if not already under the influence of another bless
        self.factory.combatant.concentration_effect = self
        for target in self.combatants:
            for mod in target.saving_throws_dice_mod.values():
                mod.append((1, 4))
            target.to_hit_dice_mod.append((1, 4))

    def deactivate(self):
        self.factory.combatant.break_concentration()
        for target in self.combatants:
            for mod in target.saving_throws_dice_mod.values():
                mod.remove((1, 4))
            target.to_hit_dice_mod.remove((1, 4))

    def deactivate_for_combatant(self, combatant):
        assert False

    def is_affecting(self, combatant):
        return combatant in self.combatants

    def calculate_threat(self, **kwargs):
        total_threat_increase = 0
        for ally in self.combatants:
            max_threat_increase = 0
            afs = get_attack_factories(ally)
            for af in afs:
                eligible_targets = af.get_eligible_targets()
                for et in eligible_targets:
                    threat_inc = af.calculate_threat_to_target_delta(et, {ThreatModifierType.TO_HIT_DIE: (1, 4)})
                    max_threat_increase = max(threat_inc, max_threat_increase)
            total_threat_increase += max_threat_increase
        return total_threat_increase * SAVING_THROW_BONUS_MULTIPLIER * ROUND_HORIZON

    def calculate_threat_for_attack(self, combatant, attack, *args, **kwargs):
        """
        Threat estimation generated by the instantiated ability.
        """
        try:
            return attack.calculate_threat_delta({ThreatModifierType.TO_HIT_DIE: (1, 4)})
        except AttributeError:
            return 0


    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        if not is_affected_by_any(self.factory.combatant, Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
            coords_for_first = set(nf.get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.combatants[0]).get(),
                distances,
                self.factory.combatant.size.value,
                BlessFactory.range,
                self.factory.combatant.id))
            coords_for_second = set(nf.get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.combatants[1]).get(),
                distances,
                self.factory.combatant.size.value,
                BlessFactory.range,
                self.factory.combatant.id))
            coords_for_third = set(nf.get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.combatants[2]).get(),
                distances,
                self.factory.combatant.size.value,
                BlessFactory.range,
                self.factory.combatant.id))
            return list(coords_for_third.intersection(coords_for_first.intersection(coords_for_second)))  # Strangely no visibility required
        elif battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.combatants[0]) <= BlessFactory.range \
            and battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.combatants[1]) <= BlessFactory.range \
            and battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.combatants[2]) <= BlessFactory.range:
            return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]
        return None

File: c++/spells/bless.cpp
#include "spells/bless.hpp"
#include "core/battle_map.hpp"
#include "core/combatant.hpp"
#include "effects/effect_tracker.hpp"
#include <algorithm>
#include <sstream>
#include <unordered_set>

namespace enc
{
  namespace
  {
    void addGroups(const std::vector<Combatant *> &targets, std::size_t groupSize, std::size_t start, std::vector<Combatant *> &current,
                   std::vector<std::vector<Combatant *>> &groups)
    {
      if(current.size() == groupSize)
        {
          groups.push_back(current);
          return;
        }
      for(std::size_t i = start; i < targets.size(); ++i)
        {
          current.push_back(targets[i]);
          addGroups(targets, groupSize, i + 1, current, groups);
          current.pop_back();
        }
    }

    double bestAttackDeltaFor(Combatant *combatant, const ThreatModifiers &mods)
    {
      double best = 0.0;
      auto collect = [&](const std::vector<std::shared_ptr<ActoidFactory>> &factories) {
        for(const auto &factory : factories)
          {
            if(auto *direct = dynamic_cast<DirectThreatFactory *>(factory.get()))
              {
                if(factory->hasFlag(FactoryFlags::IS_ATTACK_LIKE))
                  {
                    for(auto *target : BattleMap::getInstance().getNonSwallowedEnemiesWithinRadius(combatant, direct->getRange()))
                      {
                        best = std::max(best, direct->calculateThreatToTargetDelta(target, mods));
                      }
                  }
              }
          }
      };
      collect(combatant->getActionFactoriesConst());
      collect(combatant->getBonusActionFactoriesConst());
      collect(combatant->getHasteActionFactoriesConst());
      return best;
    }
  }

  BlessFactory::BlessFactory(AbilityType abilityType, Combatant *caster, Resource *resource)
      : ThreatModifierFactory("BlessFactory", "Bless", caster, abilityType), _resource(resource)
  {
    setFlag(FactoryFlags::IS_ATTACK_MODIFIER);
  }

  std::vector<std::vector<Combatant *>> BlessFactory::getEligibleTargetGroups() const
  {
    std::vector<Combatant *> targets;
    if(_combatant->getSwallower())
      {
        targets.push_back(_combatant);
      }
    else
      {
        targets = BattleMap::getInstance().getNonSwallowedAlliesWithinRadius(_combatant, static_cast<int>(BlessFactory::range));
        targets.push_back(_combatant);
      }

    std::vector<std::vector<Combatant *>> groups;
    std::vector<Combatant *> current;
    for(std::size_t groupSize = 1; groupSize <= std::min<std::size_t>(3, targets.size()); ++groupSize)
      {
        addGroups(targets, groupSize, 0, current, groups);
      }
    return groups;
  }

  std::vector<std::shared_ptr<Actoid>> BlessFactory::createAll(void *previousActionInDag)
  {
    std::vector<std::shared_ptr<Actoid>> result;
    for(auto &group : getEligibleTargetGroups())
      {
        result.push_back(std::make_shared<Bless>(std::move(group), *this));
      }
    return result;
  }

  std::shared_ptr<Actoid> BlessFactory::create(void *target)
  {
    auto *targets = static_cast<std::vector<Combatant *> *>(target);
    return std::make_shared<Bless>(*targets, *this);
  }

  double BlessFactory::calculateThreatToTarget(Combatant *target, const Kwargs &kwargs) const
  {
    ThreatModifiers mods;
    mods.set(ThreatModifierType::TO_HIT_DIE, Bless::bonusDie);
    return bestAttackDeltaFor(target, mods) * BlessFactory::SAVING_THROW_BONUS_MULTIPLIER * BlessFactory::ROUND_HORIZON;
  }

  Bless::Bless(std::vector<Combatant *> targets, BlessFactory &factory)
      : Effect(factory.getCombatant()), AttackThreatModifier(factory, ActoidFlags::IS_SPELL | ActoidFlags::IS_ATTACK_MODIFIER, factory._abilityType),
        CombatantEffect(factory.getCombatant(), std::move(targets)), LimitedDurationEffect(factory.getCombatant(), 10), _factory(factory)
  {}

  void Bless::activate(const Kwargs &kwargs)
  {
    for(auto *target : _combatants)
      {
        target->addToHitDiceMod(Bless::bonusDie);
        for(SavingThrow savingThrow : {SavingThrow::STR, SavingThrow::DEX, SavingThrow::CON, SavingThrow::INT, SavingThrow::WIS, SavingThrow::CHA})
          {
            target->addSavingThrowDiceMod(savingThrow, Bless::bonusDie);
          }
      }
    _factory.getCombatant()->setConcentrationEffect(std::dynamic_pointer_cast<Effect>(shared_from_this()));
  }

  void Bless::deactivate()
  {
    for(auto *target : _combatants)
      {
        target->removeToHitDiceMod(Bless::bonusDie);
        for(SavingThrow savingThrow : {SavingThrow::STR, SavingThrow::DEX, SavingThrow::CON, SavingThrow::INT, SavingThrow::WIS, SavingThrow::CHA})
          {
            target->removeSavingThrowDiceMod(savingThrow, Bless::bonusDie);
          }
      }
    _factory.getCombatant()->breakConcentration();
  }

  bool Bless::deactivateForCombatant(Combatant *combatant)
  {
    combatant->removeToHitDiceMod(Bless::bonusDie);
    for(SavingThrow savingThrow : {SavingThrow::STR, SavingThrow::DEX, SavingThrow::CON, SavingThrow::INT, SavingThrow::WIS, SavingThrow::CHA})
      {
        combatant->removeSavingThrowDiceMod(savingThrow, Bless::bonusDie);
      }
    _combatants.erase(std::remove(_combatants.begin(), _combatants.end(), combatant), _combatants.end());
    return !_combatants.empty();
  }

  std::string Bless::toString() const
  {
    std::ostringstream out;
    out << "Bless on ";
    for(std::size_t i = 0; i < _combatants.size(); ++i)
      {
        if(i > 0)
          {
            out << (i + 1 == _combatants.size() ? " and " : ", ");
          }
        out << _combatants[i]->_name;
      }
    return out.str();
  }

  std::string Bless::shorthandStr() const { return "Bless"; }

  double Bless::calculateThreat(const Kwargs &kwargs)
  {
    double total = 0.0;
    for(auto *target : _combatants)
      {
        total += _factory.calculateThreatToTarget(target, kwargs);
      }
    return total;
  }

  double Bless::calculateThreatForAttack(Combatant *attacker, Actoid *attack, const Kwargs &kwargs)
  {
    if(std::find(_combatants.begin(), _combatants.end(), attacker) == _combatants.end())
      {
        return 0.0;
      }
    if(auto *direct = dynamic_cast<DirectThreat *>(attack))
      {
        ThreatModifiers mods;
        mods.set(ThreatModifierType::TO_HIT_DIE, Bless::bonusDie);
        return direct->calculateThreatDelta(mods);
      }
    return 0.0;
  }

  std::optional<CoordVector> Bless::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    BattleMap &battleMap = BattleMap::getInstance();
    Combatant *caster = _factory.getCombatant();
    Coord currCoord = battleMap.getCombatantCoordinates(*caster).getRoot();
    if(caster->getSwallower())
      {
        return CoordVector{currCoord};
      }
    if(caster->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        for(auto *target : _combatants)
          {
            if(battleMap.getCartesianDistanceCombatants(*caster, *target) > static_cast<int>(BlessFactory::range))
              {
                return CoordVector{};
              }
          }
        return CoordVector{currCoord};
      }

    std::unordered_set<Coord> intersection;
    bool first = true;
    for(auto *target : _combatants)
      {
        CoordVector coords = battleMap.getFreeCoordsInCartesianRange(battleMap.getCombatantCoordinates(*target).get(), distances, caster->getSize(),
                                                                     static_cast<int>(BlessFactory::range), caster->_instanceId);
        if(first)
          {
            intersection.insert(coords.begin(), coords.end());
            first = false;
          }
        else
          {
            for(auto it = intersection.begin(); it != intersection.end();)
              {
                if(std::find(coords.begin(), coords.end(), *it) == coords.end())
                  {
                    it = intersection.erase(it);
                  }
                else
                  {
                    ++it;
                  }
              }
          }
      }
    return CoordVector(intersection.begin(), intersection.end());
  }
}
